Introducing `TaylorSolution`

PerezHz opened the pull request:
This PR introduces the `TaylorSolution` struct, a return type for `taylorinteg` output. In summary, whereas we now currently do:

```julia
t, x = taylorinteg(...)
```
the `t` and the `x` are now wrapped in a `TaylorSolution`, such that
```julia
sol = taylorinteg(...)
```
where `sol isa TaylorSolution` and we can retrieve the output time vector and solution array from `sol.t, sol.x`. Dense output is also now supported via the `dense = true` keyword, such that a solution `sol` can be called at time `t` as `sol(t)` to return the value of the solution at `t`. Here, "interpolation" occurs by simply evaluating the Taylor series polynomial at the appropriate time, thanks to time-step control ensuring the Taylor polynomial can be expanded at any time within the time-step size. In the proposed behavior, when `dense = false`, simply no dense output is produced. Besides some reworking of the TaylorIntegration internals, the DiffEq interface has also been updated, such that dense output is now also supported (we get dense output almost for free in the common interface just by adding the appropriate dispatches). ~~This PR depends on JuliaDiff/TaylorSeries.jl#355 to work properly.~~ (UPDATE: this PR no longer requires JuliaDiff/TaylorSeries.jl#355 to work properly).

A quick example how things work:

```julia
using TaylorIntegration

order = 25
abstol = 1e-20

@taylorize function kepler1!(dq, q, p, t)
    local μ = -1.0
    r_p2 = ((q[1]^2)+(q[2]^2))
    r_p3d2 = r_p2^1.5
    dq[1] = q[3]
    dq[2] = q[4]
    newtonianCoeff = μ / r_p3d2
    dq[3] = q[1] * newtonianCoeff
    dq[4] = q[2] * newtonianCoeff
    return nothing
end

q0 = [0.2, 0.0, 0.0, 3.0]

sol = taylorinteg(kepler1!, q0, 0.0, 10pi, order, abstol, dense=true, maxsteps=1000)
sol.t # vector of time-steps
sol.x # solution array
sol(1.547) # solution evaluated at t = 1.547
```

Points to solve before merging:
- [x] Adapting `TaylorSolution` for root-finding and Lyapunov spectrum `taylorinteg` specializations ~~(still in the works)~~. Update: a working version is now in place.
- [X] Agreeing on how we can add a `TaylorSolution` struct for ADS integrations (#178).
- [x] Update docs.
- [x] Fix JuliaDiff/TaylorSeries.jl#359

lbenet commented:
> ... Dense output is also now supported via the `dense = true` keyword, such that a solution `sol` can be called at time `t` as `sol(t)` to return the value of the solution at `t`. Here, "interpolation" occurs by simply evaluating the Taylor series polynomial at the appropriate time, thanks to time-step control ensuring the Taylor polynomial can be expanded at any time within the time-step size. In the proposed behavior, when `dense = false`, simply no dense output is produced...

I haven't begun reviewing this PR. Yet, from the comment motivating this PR, I think the evaluation `sol(t)` should only be performed for dense output (so an `@assert` may be needed), or we must save all intermediate Taylor polynomials also for `dense=false`. Take for instance the current implementation with (time) ranges: we do not return all intermediate steps, which is fine, because the user may not care about them; to be consistent with this, `sol(t)` should return an error. (Note that with `dense=true`, even for this example with time ranges the output should be truly the dense one.)

PerezHz commented:
> I haven't begun reviewing this PR. Yet, from the comment motivating this PR, I think the evaluation sol(t) should only be performed for dense output (so an @assert may be needed), or we must save all intermediate Taylor polynomials also for dense=false. Take for instance the current implementation with (time) ranges: we do not return all intermediate steps, which is fine, because the user may not care about them; to be consistent with this, sol(t) should return an error. (Note that with dense=true, even for this example with time ranges the output should be truly the dense one.)

Thanks for the comment! I agree this PR still has to be polished a bit. Currently there is already in place a mechanism to save the polynomial coefficients only when `dense=false`; a similar mechanism can be put in place to return an error when `sol(t)` is called `dense=false`. Otherwise, similar to what is done in the DiffEq ecosystem, a linear interpolation can be returned when `dense=false`, so that `sol(t)` still has a meaning when `dense=false`, but without saving the whole polynomial.

lbenet commented:
In an offline discussion with @LuEdRaMo, we discussed how easy can it be to adapt this proposal with the binary-tree structure involved in ADS; see #178. The crutial aspect are the fields `x` (and perhaps `p` too) of `TaylorSolution`, which are abstract vectors. 

The point I want to rise is, if `TaylorSolution` can be taylored so it also accepts the`ADSBinaryNode`, or should it (at some point, either in the implementation of #178 or as an outer constructor) convert an `ADSBinaryNode` to a vector of vectors? Or should `ADSBinaryNode` be a subtype of vector?

Thoughs are welcome...

PerezHz commented:
> The point I want to rise is, if TaylorSolution can be taylored so it also accepts theADSBinaryNode, or should it (at some point, either in the implementation of https://github.com/PerezHz/TaylorIntegration.jl/pull/178 or as an outer constructor) convert an ADSBinaryNode to a vector of vectors? Or should ADSBinaryNode be a subtype of vector?

I think this is an important point and because of this I asked a related question in #178 (see [this comment](https://github.com/PerezHz/TaylorIntegration.jl/pull/178#discussion_r1564944341)). If we can represent the solution of an ADS integration at a certain time-step as an `ADSBinaryNode`, then we can simply make the `x` field of the `TaylorSolution` a `Vector{ADSBinaryNode}` for an ADS integration. In the next few commits I was planning anyway to also rework the internals of `taylorinteg` and `TaylorSolution` to make this more transparent. And/or, we could have a `TaylorADSSolution <: AbstractTaylorSolution` as a return type of an ADS integration. With some bit of effort I think we can make the ends meet.

lbenet commented:
I like the possibility of `TaylorADSSolution`; in any case, your point is well taken: what if no ADS split is needed (or requested) in terms of `ADSBinaryNode`?

cc @LuEdRaMo

LuEdRaMo commented:
The output of an ADS integrator can be expressed as a `Vector{X}` where `X` is itself a vector of "splits", and that layout would be naturally compatible with `TaylorSolution`. However, the implementation of `ADSBinaryNode` in https://github.com/PerezHz/TaylorIntegration.jl/pull/178 is a fundamentally different structure, it is not a vector of vectors (VoV) but a binary tree. The ADS integrator could work internally with `ADSBinaryNode` and then convert the output to VoV but then, what is the point of having `ADSBinaryNode` in the first place? Therefore, from the point of view of this PR, the best thing to do would be to rewrite https://github.com/PerezHz/TaylorIntegration.jl/pull/178 to work with VoV from the beginning.

PerezHz commented:
> However, the implementation of ADSBinaryNode in https://github.com/PerezHz/TaylorIntegration.jl/pull/178 is a fundamentally different structure, it is not a vector of vectors (VoV) but a binary tree.

On a second thought, since the ADS method of `taylorinteg` returns a `ADSBinaryNode`, we can simply make `ADSBinaryNode{N, M, T} <: AbstractTaylorSolution{T, TaylorN{T}}` (and perhaps rename `ADSBinaryNode` to `TaylorADSSolution` or the like after merging of #192, but we can discuss this detail later).

PerezHz commented:
This latter possibility would allow us to preserve the tree structure for ADS, which makes sense. The only remaining point would be how to reconcile this with the DiffEq interface, which is why I was thinking in the first place how to wrap the ADS solution at a time-step as an object and write the solution as a vector of such objects. But again, we can take a look at that down the road.

PerezHz commented:
Disabled some root-finding tests temporarily related to JuliaDiff/TaylorSeries.jl#359.

PerezHz commented:
Modulo fixing of JuliaDiff/TaylorSeries.jl#359 and re-enabling the tests affected by it, this is now ready for review. Besides wrapping the solution of calls of `taylorinteg` in a `TaylorSolution`, I think the main contribution of this PR is to add support for doing JT integrations via the common interface with DiffEq, as well as avoiding some code repetitions in internal API `_taylorinteg` methods. cc: @lbenet @LuEdRaMo

lbenet reviewed (commented):
I've left a bunch of comments, some suggesting to add tests or docstrings. Aside from those, there are some other minor suggestions. I think this is almost ready to be merged, and a new minor release should follow. Yet, before that, it would be great to make some (naive) comparisons.

PerezHz commented:
Many thanks for looking into this! 

> Yet, before that, it would be great to make some (naive) comparisons.

You mean like performance comparisons (benchmarks)?

lbenet commented:
> You mean like performance comparisons (benchmarks)?

Yes, I meant benchmarks, only to have some comparisons, which are always useful...

PerezHz commented on src/integrator/taylorsolution.jl:
```diff
@@ -0,0 +1,226 @@
+    _t = constant_term(constant_term(t))  # Current time
+    tmin, tmax = minmax(sol.t[end], t0)   # Min and max time in sol
+
+    @assert tmin ≤ _t ≤ tmax "Evaluation time outside range of interpolation"
```
This `@assert` checks that the requested interpolation time lies within the integration start and end times

lbenet replied:
You are right, my mistake, I confussed `_t` with `sol.t[1]`...

PerezHz commented on src/integrator/taylorinteg.jl:
```diff
@@ -0,0 +1,465 @@
+# This file is part of the TaylorIntegration.jl package; MIT licensed
+
+# set_psol!
```
Good suggestion, thanks! It's done now.

PerezHz commented on src/integrator/taylorinteg.jl:
```diff
@@ -0,0 +1,465 @@
+end
+@inline set_psol!(::Val{false}, args...) = nothing
+
+@inline function init_psol(::Val{true}, xv::Array{U,1}, x::Taylor1{U}) where {U<:Number}
```
Docstrings added 👍

lbenet commented on src/integrator/taylorinteg.jl:
```diff
@@ -0,0 +1,490 @@
+first argument in the call signature is `Val(true)`, sets appropriate elements of argument
+`psol`. Otherwise, when the first argument in the call signature is `Val(false)`, this
+function simply returns `nothing`. Argument `psol` is the array
+where the Taylor polynomialsmassociated to the solution will be stored, corresponding to
```
```suggestion
where the Taylor polynomials associated to the solution will be stored, corresponding to
```

PerezHz commented on src/integrator/taylorsolution.jl:
```diff
@@ -0,0 +1,226 @@
+        vecsol(tevents, nevents-1),
+        matsol(xevents, nevents-1),
+        vecsol(gresids, nevents-1),
+        nothing)
```
Moved 👍

PerezHz commented on src/integrator/taylorsolution.jl:
```diff
@@ -0,0 +1,226 @@
+        nothing,
+        nothing,
+        nothing,
+        transpose(λ))
```
Moved 👍

PerezHz commented on src/common.jl:
Indeed, I think that can be pretty handy to have. I'll give this a try; thanks for the suggestion!

lbenet replied:
If it is simple to implement, I would do it; otherwise, let's open an issue an solve it in the future.

PerezHz replied:
I took a look and there are a couple of details which might not be straight-forward; I opened an issue (see #196)

PerezHz commented on test/common.jl:
```diff
@@ -9,6 +9,11 @@ import Logging: Warn
 
     local TI = TaylorIntegration
     max_iters_reached() = "Maximum number of integration steps reached; exiting.\n"
+    larger_maxiters_needed() = "Interrupted. Larger maxiters is needed. If you "*
+        "are using an integrator for non-stiff ODEs or an automatic switching "*
+        "algorithm (the default), you may want to consider using a method for "*
+        "stiff equations. See the solver pages for more details (e.g. "*
+        "https://docs.sciml.ai/DiffEqDocs/stable/solvers/ode_solve/#Stiff-Problems)."
```
Thank you for this suggestion; I also thought of doing it similarly to how you are proposing to write this string. Turns out julia1.6 somehow doesn't like it, so I had to go for a workaround to make tests pass. Btw, I hear julia1.10 will be the new LTS, so end of support for julia1.6 is nigh...

lbenet replied:
Ok; we should then leave it as it is now. Thanks anyway for checking it.

PerezHz commented on src/integrator/taylorsolution.jl:
```diff
@@ -0,0 +1,226 @@
+        ind = lastindex(sol.t) - 1
+    elseif issorted(sol.t)       # Forward integration
+        ind = searchsortedlast(sol.t, _t)
+    elseif issorted(sol.t, rev=true) # Backward integration
+        ind = searchsortedlast(sol.t, _t, rev=true)
```
Thanks; test added 👍

PerezHz commented on src/integrator/taylorsolution.jl:
```diff
@@ -0,0 +1,226 @@
+
+(sol::TaylorSolution{T, U, N})(::Val{true}, t::TT) where {T, U, N, TT<:TaylorSolutionCallingArgs{T,U}} = error("`TaylorSolution` objects computed from calls to `taylorinteg` with `dense=false` are not callable.")
+
+function (sol::TaylorSolution{T, U, 1})(::Val{false}, t::T) where {T, U}
```
Added test 👍

PerezHz commented on src/integrator/taylorsolution.jl:
```diff
@@ -0,0 +1,226 @@
+    # Evaluate sol.x[ind] at δt
+    return (sol.p[ind])(δt)::U
+end
+function (sol::TaylorSolution{T, U, 1})(::Val{false}, t::TT) where {T, U, TT<:TaylorSolutionCallingArgs{T,U}}
```
Added test

PerezHz commented on src/integrator/taylorsolution.jl:
```diff
@@ -0,0 +1,226 @@
+    # Evaluate sol.x[ind] at δt
+    return view(sol.p, ind, :)(δt)::Vector{U}
+end
+function (sol::TaylorSolution{T, U, 2})(::Val{false}, t::TT) where {T, U, TT<:TaylorSolutionCallingArgs{T,U}}
```
Added test 👍 ; thanks for catching all these!

lbenet commented:
LGTM! Thanks a lot! Have you had chance to look up the performance impact of this PR?

lbenet commented on test/common.jl:
```diff
@@ -95,10 +100,96 @@ import Logging: Warn
+        tspan = (0.0, 10pi)
+        (@test_logs (Warn, max_iters_reached()) taylorinteg(kepler1!, q0, tspan[1], tspan[2], order, abstol,
+            dense=true, maxsteps=1))
+        @time solp = taylorinteg(kepler1!, q0, tspan[1], tspan[2], order, abstol,
```
```suggestion
        solp = taylorinteg(kepler1!, q0, tspan[1], tspan[2], order, abstol,
```

lbenet commented on test/common.jl:
```diff
@@ -95,10 +100,96 @@ import Logging: Warn
+        @test isinplace(prob) == true
+        sol1 = solve(prob, TaylorMethod(order), abstol=abstol, maxiters=2000, parse_eqs=false)
+        (@test_logs (Warn, larger_maxiters_needed()) solve(prob, TaylorMethod(order), abstol=abstol, maxiters=1))
+        @time sol2 = solve(prob, TaylorMethod(order), abstol=abstol, maxiters=2000)
```
```suggestion
        sol2 = solve(prob, TaylorMethod(order), abstol=abstol, maxiters=2000)
```

lbenet commented on test/common.jl:
```diff
@@ -95,10 +100,96 @@ import Logging: Warn
+
+        (@test_logs (Warn, max_iters_reached()) taylorinteg(kepler1!, q0TN, tspan[1], tspan[2], order, abstol,
+            dense = true, maxsteps=1))
+        @time solTN = taylorinteg(kepler1!, q0TN, tspan[1], tspan[2], order, abstol,
```
```suggestion
        solTN = taylorinteg(kepler1!, q0TN, tspan[1], tspan[2], order, abstol,
```

lbenet reviewed (commented):
I just left 4 suggestions (remove `@time` from some tests). Aside from that, I think we should simply bump the minor version, merge and register

PerezHz commented on src/integrator/taylorinteg.jl:
```diff
@@ -0,0 +1,465 @@
+@inline init_psol(::Val{false}, ::Array{U,2}, ::Array{Taylor1{U},1}) where {U<:Number} = nothing
+
+# taylorinteg
+function taylorinteg(f, x0::U, t0::T, tmax::T, order::Int, abstol::T, params = nothing;
```
The default is now `dense=true`

PerezHz commented:
> I just left 4 suggestions (remove @time from some tests). 

Thanks, I added those to run some checks; I've removed them now.

> Aside from that, I think we should simply bump the minor version, merge and register

Just as a way to help users, I would go actually for bumping the patch version with deprecations (I went ahead and added some in src/deprecated.jl as well as a `depwarn` in the `TaylorSolution` inner constructor). The caveat is that, since there is a warning and we test the warning logs, tests will fail. I would afterwards propose release the new minor version (v0.16) with the deprecations removed (I can take care of both releases).

lbenet commented:
> Just as a way to help users, I would go actually for bumping the patch version with deprecations (I went ahead and added some in src/deprecated.jl as well as a `depwarn` in the `TaylorSolution` inner constructor). The caveat is that, since there is a warning and we test the warning logs, tests will fail. I would afterwards propose release the new minor version (v0.16) with the deprecations removed (I can take care of both releases).

As you prefer; I though about the minor version because it is a (quite) breaking release, which is not backward compatible (in the sense of the returned types/content of, say, `taylorinteg`...

lbenet commented:
I'm trying now to have some very naive benchmarks...

lbenet commented:
I run (two) extremely naive benchmarks, based on the time reported by the tests for jet transport and taylorize (second run of `include`ing the tested file), comparing this PR (with d3605313) and v0.15.3 (`main`). It seems there is not a large performance hit due to this PR, and perhaps even better performance in taylorize tests, but these statements are based on *very* naive benchmarks...

PerezHz commented:
I'm also running some naive timings on 3000 steps of the Kepler problem with 2nd-order jet transport:

```julia
varorder = 2 #the order of the variational expansion
p = set_variables("ξ", numvars=4, order=varorder) #TaylorN setup
q0 = [0.2, 0.0, 0.0, 3.0]
q0TN = q0 + p # JT initial condition
t0 = 0.0
tf = (2π)*20

@taylorize function kepler1!(dq, q, p, t)
    local μ = -1.0
    r_p2 = ( (q[1]^2) + (q[2]^2) )
    r_p3d2 = r_p2^1.5
    dq[1] = q[3]
    dq[2] = q[4]
    newtonianCoeff = μ / r_p3d2
    dq[3] = q[1] * newtonianCoeff
    dq[4] = q[2] * newtonianCoeff
    return nothing
end
```

On `main`:

```julia
@time taylorinteg(kepler1!, q0TN, t0, tf, 25, 1e-20, Val(true), maxsteps=3000);
# 4.382222 seconds (13.29 M allocations: 1.001 GiB, 4.25% gc time)
```

On `jp/solution`:
```julia
@time taylorinteg(kepler1!, q0TN, t0, tf, 25, 1e-20, maxsteps=3000, dense=true);
# 2.970461 seconds (12.05 M allocations: 989.286 MiB, 8.68% gc time)
```

Does this look more or less similar to what you're seeing?

lbenet commented:
Yes, they seem consistent wrt to the elapsed time; I'm not checking allocations, but I can guess that they reproduce your results. So, aside from the depration message, I am in favor of merging and registering the new version! Thanks truly a lot!

PerezHz commented:
> As you prefer; I though about the minor version because it is a (quite) breaking release, which is not backward compatible (in the sense of the returned types/content of, say, taylorinteg...

Ah, you're right, I've not managed to catch all the deprecations so indeed if we release a patch version like this dependant code will break. The clean way would be to catch all deprecated behavior, but this is something I overlooked at the beginning of this PR and to do it now would be a bit of a hassle. So indeed perhaps it'd be better to go straight for the minor version, without deprecations (and I'll add a breaking changes section to the release notes).

lbenet commented:
So, let's go for the minor version patch then!!
